src/encryption.rs

```diff
@@ -0,0 +1,729 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+pub use crocksdb_ffi::{
+    self, DBEncryptionKeyManagerInstance, DBEncryptionMethod, DBFileEncryptionInfo,
+};
+
+use libc::{c_char, c_void, size_t};
+use std::ffi::{CStr, CString};
+use std::fmt::{self, Debug, Formatter};
+use std::io::Result;
+use std::ptr;
+use std::sync::Arc;
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct FileEncryptionInfo {
+    pub method: DBEncryptionMethod,
+    pub key: Vec<u8>,
+    pub iv: Vec<u8>,
+}
+
+impl Default for FileEncryptionInfo {
+    fn default() -> Self {
+        FileEncryptionInfo {
+            method: DBEncryptionMethod::Unknown,
+            key: vec![],
+            iv: vec![],
+        }
+    }
+}
+
+impl Debug for FileEncryptionInfo {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "FileEncryptionInfo [method={}, key=...<{} bytes>, iv=...<{} bytes>]",
+            self.method,
+            self.key.len(),
+            self.iv.len()
+        )
+    }
+}
+
+impl FileEncryptionInfo {
+    pub unsafe fn copy_to(&self, file_info: *mut DBFileEncryptionInfo) {
+        crocksdb_ffi::crocksdb_file_encryption_info_set_method(file_info, self.method);
+        crocksdb_ffi::crocksdb_file_encryption_info_set_key(
+            file_info,
+            self.key.as_ptr() as *const c_char,
+            self.key.len() as size_t,
+        );
+        crocksdb_ffi::crocksdb_file_encryption_info_set_iv(
+            file_info,
+            self.iv.as_ptr() as *const c_char,
+            self.iv.len() as size_t,
+        );
+    }
+}
+
+pub trait EncryptionKeyManager: Sync + Send {
+    fn get_file(&self, fname: &str) -> Result<FileEncryptionInfo>;
+    fn new_file(&self, fname: &str) -> Result<FileEncryptionInfo>;
+    fn delete_file(&self, fname: &str) -> Result<()>;
+    fn link_file(&self, src_fname: &str, dst_fname: &str) -> Result<()>;
+    fn rename_file(&self, src_fname: &str, dst_fname: &str) -> Result<()>;
+}
+
+// Copy rust-owned error message to C-owned string. Caller is responsible to delete the result.
+fn copy_error<T: Into<Vec<u8>>>(err: T) -> *const c_char {
+    let cstr = CString::new(err).unwrap();
+    unsafe { libc::strdup(cstr.as_ptr()) }
```

Comment:
```suggestion
    CString::new(err).into_raw()
```

To save an extra allocation, we can just use `CString::into_raw()`.

Reply:
We cannot. The document of `CString::into_raw()` says the result is not safe to be freed in C.